Introduce gitflow-incremental-builder

famod opened the pull request:
Relates to #11622
Fixes #6260

Note: master, release branches (e.g. 1.10) and backport-branches ~~(without PR)~~ are not built incrementally (yet)!

Some general remarks:
- various details have been discussed here: https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/gitflow-incremental-builder.20(GIB)
- for branches without a PR, I decided to use `quarkusio/quarkus master` and not the fork's `master` as the reference because no one on Zulip had a (strong) opinion about this and since some (most?) devs don't seem to push to their fork's `master` at all, fetching `quarkusio/quarkus master` seemed like the best compromise
- `fetch-depth: 0` is required to have a fully usable history and the reference branch available (in case the reference is _not_ `quarkusio/quarkus master`, which is fetched explicitely)
- the Maven profile is set up in a way that allows local usage as well, with a slightly different behaviour than in CI - e.g. I usually do want GIB to care about uncommitted and untracked changes locally, but not in CI
- GIB cannot do anything about the gradle jobs, so those will always perform full builds
- I tested this rather thoroughly but in my quarkus fork I could not test the real setup. So in the worst case (in case of errors after merging this) I/we have to be quick to fix things or revert this.

I'll also add some review comments myself.

_Edit:_ List of **TODOs** for a better overview (updated on demand):
- [x] never enable GIB for backport branches, even if there is a PR (that defines the target branch) - see: https://github.com/quarkusio/quarkus/pull/13243#discussion_r524345456
- [x] wait for #13815 to fix [`ServerSerialisers` failure in `micrometer-prometheus` integration tests](https://github.com/quarkusio/quarkus/pull/13243#issuecomment-735000452)
- [x] discuss/consider renaming Maven profile from `gib` to `incremental` or similar
- [x] add GIB to Dependabot
- [x] add documentation to `CONTRIBUTING.md` (as discussed on the mailing list)
- [x] wait for #13218 because of the [descriptor dependency problem](https://github.com/quarkusio/quarkus/pull/13243#issuecomment-725946572)
  - [x] add defaultly enabled profile for `devtools/bom-descriptor-json` that can be disabled in CI to avoid executing all ITs if e.g. only a single extension was changed
  - [x] finalize test with multiple forks (see _Idea for testing this closer to reality_ further down)
- [x] make sure that `independent-projects` are run in case `.github/**` or `.mvn/**` is changed
- [x] introduce temporary "CI shadowing" [as requested](https://github.com/quarkusio/quarkus/pull/13243#issuecomment-777650957)
- [x] remove "forces build of" comments (stale)

**Idea for testing this closer to reality:**
Someone with write access to `quarkusio/quarkus` could create a branch named e.g. `gib-test` in `quarkusio/quarkus` from this PR branch and could then branch off `gib-test` in his/her fork (!) and create a test PR.

famod commented on .github/workflows/ci-actions.yml:
```diff
@@ -83,7 +83,7 @@ jobs:
       - name: Build
         run: |
-          mvn -e -B -DskipTests -DskipITs -Dno-format clean install
+          mvn -e -B -DskipTests -DskipITs -Dinvoker.skip -Dno-format -Dgib -Dgib.disable clean install
```
Note:
- `-Dinvoker.skip` is offtopic but still worthwhile (skip maven invoker tests, e.g. in enforcer-rules)
- `-Dgib -Dgib.disable` just triggers the download of GIB and it's dependencies

gastaldi replied:
I wonder if we should replace `DskipTests -DskipITs -Dno-format` with `-Dquickly`

famod replied:
IIRC, a few weeks ago someone said: `-Dquickly` is only for local usage.
But yeah, I also do see your point.

aloubyansky replied:
-Dquickly skips the formatting check though. There seem to be tests that use maven invoker, won't they be affected by `-Dinvoker.skip`?

famod replied:
> There seem to be tests that use maven invoker, won't they be affected by `-Dinvoker.skip`?

Yes, that's what should happen. This initial JDK11 build shall only _build_ everything and populate a local Maven repo that is then reused by subsequent jobs. This initial job shall not execute _any_ tests (see also `-DskipTests`), AFAICS.

gsmet replied:
Please do not use `-Dquickly` in CI build.

This thing will evolve depending on what people need for local usage and we don't want CI to be impacted.

famod replied:
Btw, the docs job _is_ using `-Dquickly`: https://github.com/quarkusio/quarkus/blob/master/.github/workflows/doc-build.yml#L45

famod commented on .github/workflows/ci-actions.yml:
```diff
@@ -95,6 +95,30 @@ jobs:
       - name: Delete Local Artifacts From Cache
         shell: bash
         run: rm -r ~/.m2/repository/io/quarkus
+      - name: Get GIB arguments
```
I was looking long and hard for a possibility to have some kind of conditional logic in the `env` section but it seems this is not supported by GH Workflows.
That's why I ended up with this build step that creates an output that is then exposed as a job output to subsequent jobs.
I could have created a spearate job for this but then I would need to extend `needs` of each job.

famod commented on .github/workflows/ci-actions.yml:
```diff
@@ -130,6 +154,11 @@ jobs:
+          fetch-depth: 0
+      - name: Add quarkusio remote
+        shell: bash
+        run: git remote add quarkusio https://github.com/quarkusio/quarkus.git
```
Notes:
- this entire block is now present for almost each job
- I decided to only add the remote, not directly fetch from it, so that GIB can decide whether or not to fetch anything at all (see the "Get GIB arguments" step further up for the cases where GIB disables itself)

famod commented on .github/workflows/ci-actions.yml:
```diff
@@ -154,7 +183,8 @@ jobs:
         shell: bash
         run: tar -xzf maven-repo.tgz -C ~
       - name: Build with Maven
-        run: eval mvn $JVM_TEST_MAVEN_OPTS install -pl !integration-tests/gradle -pl !integration-tests/maven -pl !integration-tests/devtools ${{ matrix.java.maven_args}}
```
I had some nasty formatting/escaping problems with `eval`, especially with `-Dgib.disableIfBranchRegex='master|\d+\.\d+|.*backport.*'` (see further up).
AFAICT `eval` ist not required so I simply removed it from every `mvn` call.

famod commented on .github/workflows/ci-actions.yml:
```diff
@@ -452,9 +512,9 @@ jobs:
+        run: mvn -B --settings .github/mvn-settings.xml -Dno-native -Dno-format -DskipTests -Dtcks install ${{ needs.build-jdk11.outputs.gib_args }}
       - name: Verify with Maven
-        run: mvn -B --settings .github/mvn-settings.xml -f tcks/pom.xml verify
+        run: mvn -B --settings .github/mvn-settings.xml -Dtcks -pl tcks -amd verify ${{ needs.build-jdk11.outputs.gib_args }}
```
While `-f` is more efficient, here with we need `-pl` we need the entire reactor (=all modules) so that GIB can properly detect changes and map them to modules.
With `-f` GIB would only "see" a fraction of the modules and change detection would not work as we need it to.

gsmet replied:
Won't this thing run the build and the tests of all the modules triggered by amd?

famod replied:
No, because the profile overwrites the modules so that only `tcks` is left:
```xml
        <profile>
            <id>tcks</id>
            <activation>
                <property>
                    <name>tcks</name>
                </property>
            </activation>
            <modules>
                <module>tcks</module>
            </modules>
        </profile>
```

gsmet replied:
Hmmm, I do that in HV and it adds the modules, it doesn't overwrite them.

famod replied:
You are absolutely right!
The thing here is that `-pl tcks -amd` builds `Quarkus - TCK - Parent` and everything that depends on it, which is basically just all those tcks submodules.

famod replied:
...and from those modules GIB is removing the ones that are not changed (or don't depend on changed modules).

Example:
```
$ mvn -Dgib -Dgib.dbc -Dtcks -pl tcks -amd validate -Dgib.dsph
[INFO] Scanning for projects...
[INFO] gitflow-incremental-builder 3.12.0 starting...
[INFO] ------------------------------------------------------------------------
[INFO] Changed Artifacts:
[INFO]
[INFO] quarkus-vertx (but deselected)
[INFO]
[INFO] gitflow-incremental-builder exiting...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO]
[INFO] Quarkus - TCK - MicroProfile Health                                [jar]
[INFO] Quarkus - TCK - MicroProfile JWT                                   [jar]
[INFO] Quarkus - TCK - MicroProfile Reactive Messaging                    [jar]
[INFO] Quarkus - TCK - MicroProfile OpenAPI                               [jar]
```
PS: I'm lazy 😉; `dbc` is `disableBranchComparison` and `dsph` is `disableSelectedProjectsHandling`.

famod replied:
This should also be resolved by now.

famod commented on .github/workflows/ci-actions.yml:
```diff
@@ -512,14 +577,13 @@ jobs:
-          do modules+=("integration-tests/$i"); done
-          IFS=,
-          eval mvn -pl "${modules[*]}" $NATIVE_TEST_MAVEN_OPTS
+          do modules+="integration-tests/$i,"; done
```
Additionally to `eval` (see futher up) I also had issues with `IFS=,`.
Since Maven doesn't mind a trailing `,`, I simplified it all a bit.

gastaldi commented on pom.xml:
```diff
@@ -205,6 +206,36 @@
             </modules>
         </profile>
+        <profile>
+            <id>gib</id>
```
I am not totally certain on this, but can we name it `incremental` (and the activation property)? I think it's more intuitive (although `gib` is shorter to type, but I think readability is important)

The reason why I am, uncertain is that I see that you have other parameters when you use `-Dgib` that also starts with `-Dgib.`, so maybe it makes sense.

I'm curious to know what others think.

famod replied:
Yeah sure, why not.

famod replied:
@aloubyansky @gsmet @maxandersen what would you prefer? `-Dgib` or `-Dincremental`?
All other `gib.*` properties must stay this way since those are GIB built-ins.

maxandersen replied:
"-Dmore-quickly" ? :) 

just kidding.

"-Dincremental" works for me. better to have profile be about the intent rather than the technology used.

famod replied:
Ok, renamed to `incremental`.

gastaldi commented on pom.xml:
```diff
@@ -35,6 +35,7 @@
         <postgres.url>jdbc:postgresql:hibernate_orm_test</postgres.url>
 
         <nexus-staging-maven-plugin.version>1.6.8</nexus-staging-maven-plugin.version>
+        <gitflow-incremental-builder.version>3.12.0</gitflow-incremental-builder.version>
```
Add to the Dependabot descriptor?

famod replied:
Done.

famod commented:
It just occured to me that if we use this locally, we will most likely run into a problem similar to this: https://github.com/quarkusio/quarkus/pull/13218#issuecomment-725484333

Of course this PR here is not about parallel builds, but incremental builds are also affected because in case an extension changes (or is added) in a way that is relevant for the platform descriptor generation, an incremetal build will _not_ include `quarkus-bom-quarkus-platform-descriptor` since that module has no dependencies at all.

In CI we are safe since we have this initial single threaded build that will produce a consistent platform descriptor.
Therefore I don't think this problem should be blocking this PR, but we maybe should _not_ advertise local usage just yet.

PS: I _might_ be able to work around this via/in GIB but then again we also have local builds without GIB, so we need a general solution.

aloubyansky commented:
The descriptor in fact does depend on the presence of the extensions it includes. So we could introduce dependencies at least on the POMs of all the extensions.

famod commented:
> The descriptor in fact does depend on the presence of the extensions it includes. So we could introduce dependencies at least on the POMs of all the extensions.

Yeah, that would be the hideously verbose but standard Maven solution. I'd like to continue this discussion in #13218.

gsmet commented on .github/workflows/ci-actions.yml:
```diff
@@ -95,6 +95,30 @@ jobs:
+          GIB_ARGS="-Dgib -Dgib.disableSelectedProjectsHandling -Dgib.untracked=false -Dgib.uncommitted=false"
+          if [ -n "$PULL_REQUEST_BASE" ]
+          then
+              # The PR defines a clear merge target so just use that branch for reference
```
When does it happen exactly? I would like to understand when this case happens and when the other case happens?

Typically, I don't understand why we only disable backport PRs in the other case.

gsmet replied:
Also one question related to that: I don't understand exactly how things work for a PR because the remote master might be totally different from the one used to create the PR (if not rebased). I fail to understand how it works in this case?

famod replied:
> When does it happen exactly? I would like to understand when this case happens and when the other case happens?

As far I was able to observe/test, this case happens when the workflow run was triggered by a pull request and not just merely by a push to a branch that does not yet have a PR.

> Typically, I don't understand why we only disable backport PRs in the other case.

The idea here is that there is no guessing when you have a PR because the PR clearly defines the target branch. So for some `foo-backport-bar` branch (or whatever) that already has a PR, GIB can simply take the target branch from the PR as reference and compare to that (e.g. `1.10`).

>  I don't understand exactly how things work for a PR because the remote master might be totally different from the one used to create the PR 

When gathering the changes, the source branch (=from where the "working branch", or in GIB terms "base branch", was branched off) is not so important. GIB or better to say JGit will find the common [merge base](https://git-scm.com/docs/git-merge-base) and do the comparison from there.

Btw, I am still hoping that someone with write access to this repo will try my "Idea for testing this closer to reality" (see PR description).
In my fork I wasn't able to properly test with multiple repos. But come to think of it, maybe I should just fork again from my fork to try to simulate it.

famod replied:
PS, regarding:
> Typically, I don't understand why we only disable backport PRs in the other case.

I've just corrected the following bit in the PR description which I think contributed to the confusion:

From:
> Note: master, release branches (e.g. 1.10) and backport-PR-branches are not built incrementally (yet)!

To:
> Note: master, release branches (e.g. 1.10) and backport-branches (without PR) are not built incrementally (yet)!

famod replied:
PPS:
> But come to think of it, maybe I should just fork again from my fork to try to simulate it.

Hm, doesn't work. You can only have one fork.

gsmet replied:
So... maybe I'm missing something but from what I understood of what you say...

`if [ -n "$PULL_REQUEST_BASE" ]`

We are in the pull request case. We have a proper sha to which we can compare our changes. `PULL_REQUEST_BASE` is not a branch but a sha that points to the commit on top of which we have added the commits in the PR? Right?

If so, in this case, I don't want to use GIB for backport PRs either. I remember having said that before. In the case of the backport PRs, I want a full build in all cases.

`else`

If this case doesn't concern PRs, I would just disable the incremental build altogether. Especially since things are not very clear and we don't have a sha to compare to.

famod replied:
> `PULL_REQUEST_BASE` is not a branch but a sha

Yes.

> a sha that points to the commit on top of which we have added the commits in the PR? Right?

Only if e.g. `master` did not receive any new commits since the PR branch was created.
There is a good explanation here: https://github.community/t/how-to-get-base-sha-when-using-pull-request-target/142021/6

> If so, in this case, I don't want to use GIB for backport PRs either.

Personally I think with a PR it is pretty safe already since you have a clear merge target/comparison reference (the most recent SHA at PR creation time from e.g. `1.10`). Additionally, with this PR, `master` and release branches are still built fully.
But these are just my 2 cents and if you want to play it extra safe so be it.

> If this case doesn't concern PRs, I would just disable the incremental build altogether. 

That would mean that branches in forks are always full builds. That's a pretty "hard" decision, IMO.
That's also why I asked on Zulip whether we can always expected "backport" to be contained in names of branch that do not target `master` but a release branch instead (e.g. `1.10`).

Btw, it is not a problem if `master` has advanced since creating the branch because the history of `master` is never rewritten (I suppose).

famod replied:
Update: `*.backport.*` branches are now _not_ built incrementally, even if there is a PR.

famod replied:
I think is resolved. As of now, branches in forks will also be built incrementally (which I definitely recommend) and `*.backport.*` branches are never built incrementally, with or without PR.

gsmet reviewed (changes requested):
I added a few questions.

famod commented:
Weird, I cannot comment directly on this question: https://github.com/quarkusio/quarkus/pull/13243#discussion_r523061097

> Won't this thing run the build and the tests of all the modules triggered by amd?

@gsmet No, because the profile overwrites the modules so that only `tcks` is left:
```xml
        <profile>
            <id>tcks</id>
            <activation>
                <property>
                    <name>tcks</name>
                </property>
            </activation>
            <modules>
                <module>tcks</module>
            </modules>
        </profile>
```

famod commented:
Rebased & updated:
> - [x] never enable GIB for backport branches, even if there is a PR (that defines the target branch) - see: https://github.com/quarkusio/quarkus/pull/13243#discussion_r524345456
> - [x] add GIB to Dependabot

famod commented:
I'll continue with this after #13218 was merged.

famod commented:
I've just rebased and had to adapt to https://github.com/quarkusio/quarkus/commit/5244c9cbe47287718b045fdd43f1e27b93d0c643.
Since that commit brought in the "external" repo `quarkus-rest-testsuite` (and a `mvn` invocation for it), `GIB` is not able to help and therefore this step will not be run incrementally (takes ~4mins currently).
/cc @FroMage @geoand

geoand commented:
Thanks @famod

famod commented:
`PrometheusMetricsRegistryTest` fails in all "JVM Tests" jobs and I don't understand why:
```
java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    io/quarkus/resteasy/reactive/server/deployment/QuarkusRestProcessor.setupEndpoints(Lio/quarkus/deployment/Capabilities;Lio/quarkus/arc/deployment/BeanArchiveIndexBuildItem;Lio/quarkus/arc/deployment/BeanContainerBuildItem;Lio/quarkus/resteasy/reactive/common/runtime/QuarkusRestConfig;Ljava/util/Optional;Lio/quarkus/deployment/annotations/BuildProducer;Lio/quarkus/deployment/annotations/BuildProducer;Lio/quarkus/rest/server/runtime/QuarkusRestRecorder;Lio/quarkus/deployment/recording/RecorderContext;Lio/quarkus/deployment/builditem/ShutdownContextBuildItem;Lio/quarkus/vertx/http/runtime/HttpBuildTimeConfig;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/Optional;Ljava/util/Optional;Lio/quarkus/deployment/annotations/BuildProducer;Lio/quarkus/deployment/annotations/BuildProducer;Lio/quarkus/deployment/annotations/BuildProducer;Lio/quarkus/resteasy/reactive/common/deployment/ApplicationResultBuildItem;)V @1682: invokestatic
  Reason:
    Type 'org/jboss/resteasy/reactive/server/core/ServerSerialisers' (current frame, stack[6]) is not assignable to 'org/jboss/resteasy/reactive/common/core/Serialisers'
```

@ebullient do you have a clue what's going on? Thanks!

ebullient commented:
> `PrometheusMetricsRegistryTest` fails in all "JVM Tests" jobs and I don't understand why:
> 
> ```
> [ERROR] Tests run: 8, Failures: 0, Errors: 1, Skipped: 7, Time elapsed: 6.509 s <<< FAILURE! - in io.quarkus.it.micrometer.prometheus.PrometheusMetricsRegistryTest
> [ERROR] io.quarkus.it.micrometer.prometheus.PrometheusMetricsRegistryTest.testRegistryInjection  Time elapsed: 0.01 s  <<< ERROR!
> java.lang.RuntimeException: 
> java.lang.VerifyError: Bad type on operand stack
> Exception Details:
>   Location:
>     io/quarkus/resteasy/reactive/server/deployment/QuarkusRestProcessor.setupEndpoints(Lio/quarkus/deployment/Capabilities;Lio/quarkus/arc/deployment/BeanArchiveIndexBuildItem;Lio/quarkus/arc/deployment/BeanContainerBuildItem;Lio/quarkus/resteasy/reactive/common/runtime/QuarkusRestConfig;Ljava/util/Optional;Lio/quarkus/deployment/annotations/BuildProducer;Lio/quarkus/deployment/annotations/BuildProducer;Lio/quarkus/rest/server/runtime/QuarkusRestRecorder;Lio/quarkus/deployment/recording/RecorderContext;Lio/quarkus/deployment/builditem/ShutdownContextBuildItem;Lio/quarkus/vertx/http/runtime/HttpBuildTimeConfig;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/Optional;Ljava/util/Optional;Lio/quarkus/deployment/annotations/BuildProducer;Lio/quarkus/deployment/annotations/BuildProducer;Lio/quarkus/deployment/annotations/BuildProducer;Lio/quarkus/resteasy/reactive/common/deployment/ApplicationResultBuildItem;)V @1678: invokestatic
>   Reason:
>     Type 'org/jboss/resteasy/reactive/server/core/ServerSerialisers' (current frame, stack[6]) is not assignable to 'org/jboss/resteasy/reactive/common/core/Serialisers'
> ```
> 
> @ebullient do you have a clue what's going on? Thanks!

Hmm. I don't know off the top of my head. The other micrometer tests are ok.. it's just that one?

famod commented:
> The other micrometer tests are ok.. it's just that one?

The following tests are ok:
- io.quarkus.it.micrometer.jmx.JmxMetricsRegistryTest
- io.quarkus.it.micrometer.mpmetrics.MPMetricsTest
- io.quarkus.it.micrometer.native_mode.NativeMeterRegistriesTest

But in the module `micrometer-prometheus` there are no other tests, only the failing `PrometheusMetricsRegistryTest`.

ebullient commented:
> The following tests are ok:
> 
>     * io.quarkus.it.micrometer.jmx.JmxMetricsRegistryTest
>     * io.quarkus.it.micrometer.mpmetrics.MPMetricsTest
>     * io.quarkus.it.micrometer.native_mode.NativeMeterRegistriesTest
> 
> But in the module `micrometer-prometheus` there are no other tests, only the failing `PrometheusMetricsRegistryTest`.

It was the other buckets I meant (they all set up the same registries, more or less. The JMX test does not use Resteasy, but the MP Metrics test does. The projects and tests between MP Metrics and Prometheus are pretty similar in structure, so I don't understand why just the Prometheus test would fail.. and I have no idea why one ServerSerialiser would be selected over the other ... (or what the incremental build touches that would be relevant here, as it clearly all passes in the full builds.. )

famod commented:
https://github.com/vackosar/gitflow-incremental-builder#gibbuildall works around the problem but is not very efficient for 700+ modules. Still digging...

famod commented:
I found a much better alternative to `buildAll` (but it's still a workaround):
`-Dgib.forceBuildModules=resteasy-reactive` (see also: https://github.com/vackosar/gitflow-incremental-builder#gibforcebuildmodules)
This will always trigger a build of `independent-projects/resteasy-reactive/server/runtime` which contains `ServerSerialisers`.

What I don't get is why is a rebuild of this module even needed? The module is already built in the "Initial JDK 11 Build" and is be shared via the "persisted" Maven repo.
The parameters for `mvn` in that initial job are a bit different than the parameters in the JVM tests jobs but nothing that really caught my eye, except for `--settings .github/mvn-settings.xml` which is not set for the initial job.
Adding this settings file to the initial job does not change anything, though.

@geoand @FroMage any idea what's going on? I'm talking about https://github.com/quarkusio/quarkus/pull/13243#issuecomment-735000452 and subsequent comments.

maxandersen commented:
I've created https://github.com/quarkusio/quarkus/tree/gib-test that is this branch but in quarkusio ..now just need to find a PR to try with ;)

geoand commented:
> @geoand @FroMage any idea what's going on? I'm talking about [#13243 (comment)](https://github.com/quarkusio/quarkus/pull/13243#issuecomment-735000452) and subsequent comments.

I don't see why this would cause a problem :(.

famod commented:
FTR, further troubleshooting currently takes place in #13543.

famod commented:
Now that #13815  is merged, I was able to remove the `forceBuildModules` workaround (for the `ServerSerializers` problem).

PS: Still waiting for #13218 to be finalized and merged.

famod commented on .github/workflows/ci-actions.yml:
```diff
@@ -102,6 +102,33 @@ jobs:
+          # Common GIB_ARGS for all CI cases (hint: see also pom.xml):
+          # - disableSelectedProjectsHandling: required to detect changes in jobs that use -pl
+          # - untracked: to ignore files created by jobs (and uncommitted to be consistent)
+          # - forces build of quarkus-core.* (see https://github.com/quarkusio/quarkus/pull/13543#issuecomment-735955489)
```
This and the next line need to be removed (was a temporary workaround).

famod replied:
Done.

famod commented:
@n1hility I'd love to hear your opinion about this! 🙂

gsmet commented:
@famod this needs a rebase as I merged your other PR (I think that's the one that created the conflict)

So... my take on this is that we should have some shadowing of the current CI for ~ a week to be able to see what's going on. So, I would probably duplicate the workflow first, have it run for a week and then decide depending on the result of it.

We will pay a CI price for a week but if GIB works great, it should be sustainable.

We don't update the ci-actions workflow that often now that the native test list is in a JSON file so I think it should be doable without too much pain.

@famod WDYT about this strategy?

famod commented:
Thanks for your feedback, @gsmet!

> @famod WDYT about this strategy?

So the idea is to have a copy of the pre-GIB state of `ci-actions.yml` as...say... `ci-actions-not-incremental.yml` or so?
If the available resources can handle it and we mange to invest some "monitoring time", then this sounds like a good idea!
If it's too much then maybe only pick some very critical jobs? (e.g. leave out one or two JDK test jobs)

In the beginning, I actually aimed at making just a few jobs incremental as a start, but with that approach you wouldn't have a full side-by-side comparison.

IIRC, @ebullient mentioned that in a past project they had a permanent weekly (?) full build to make people more confident with the incremental approach. This might also be something to consider (additionally).

**But before we merge this I have to complete the remaining tasks.**

The one task with the new generated dependencies of `devtools/bom-descriptor-json` (#13218) is not so trivial.
If we don't do anything about it, a single change to any of the `extensions` modules will trigger _all_ `integration-tests` modules because all (most?) IT modules depend on the descriptor and thus on all extensions.
This would be a massive setback for the incremenal build efficiency.

A simplistic solution would be to take `devtools/bom-descriptor-json` out of the reactor in IT jobs via a Maven profile (`-pl !...` doesn't work here). But this means that changes to `bom-descriptor-json` or `quarkus-platform-descriptor-json-plugin` would _not_ trigger any ITs (and to some extend some of the plugin's upstream dependencies wouldn't trigger them either, dependening on the actual dependency graph).

The alternative to this would be a feature in GIB that can remove certain dependency subpaths from the reactor when certain conditions are met. Doesn't seem easy to implement, especially given the API Maven provides for this.

Next best (but slightly weird) solution might be to disable GIB in case certain modules are changed. This is something CI would have to figure out.

/cc @aloubyansky What's your take on this?

PS: Sorry for the long text but this must be done right. CI/build consistency over efficiency.

famod commented:
I just had another idea reagarding the descriptor dependency graph problem:

If we move all those generated extension dependencies to a profile (that is active by default), we could disable it for incremental builds. So changes to the descriptor, its plugin or any of the plugin's upstream dependencies would trigger all ITs, _but not a change to a single extension module_!

/cc @aloubyansky ^ (for now this feels like the best solution but maybe I'm missing something)

gsmet commented:
> So the idea is to have a copy of the pre-GIB state of ci-actions.yml as...say... ci-actions-not-incremental.yml or so?
If the available resources can handle it and we mange to invest some "monitoring time", then this sounds like a good idea!
If it's too much then maybe only pick some very critical jobs? (e.g. leave out one or two JDK test jobs)

I would better create yours as a new one for now. And, yes, we could enable only one JDK for the incremental build.

As for CI resources, we will see, it will for sure slow things down but I'm not comfortable merging this without having some sort of shadowing for a while.

> IIRC, @ebullient mentioned that in a past project they had a permanent weekly (?) full build to make people more confident with the incremental approach. This might also be something to consider (additionally).

I agree.

As for the rest, I will let you discuss it with @aloubyansky :).

famod commented:
I've rebased and added another commit that shows my idea with the depdencies profile in `bom-descriptor-json`.
If we agree on this approach, I'll finalize the indentation logic in `update-extension-dependencies.sh`.

aloubyansky commented:
> PS: Just as a recap: The initial JDK build will always install a consistent descriptor artifact and the repo is shared with the subsequent jobs. That's why you _could_ skip the descriptor module entirely in the IT jobs etc.

That is true.

famod commented:
I think I will use the `incremental` property to control that new profile in `bom-descriptor-json` instead of adding that other new clunky property.
Also, I'll test skipping `maven-install-plugin` on `incremental` in that module to avoid an inconsistent descriptor being installed (hoping that the descriptor lookup in the ITs doesn't barf and is resorting to the descriptor in the local repo).

There might also be a problem with `independent-projects` not being rebuilt if e.g. `.github/**` contains changes, because those modules don't depend on the root pom.xml (which GIB is normally assigning those changes to).
If testing confirms this (90% sure), I'll have to introduce something in GIB to support this special case. Doing this in the workflow instead would be rather verbose and fragile.

So, it doesn't get boring! 😉 But better find the problems now than later.

famod commented:
> There might also be a problem with `independent-projects` not being rebuilt if e.g. `.github/**` contains

Confirmed, as expected. I'll add `disableIfChangedRegex` (or similar) to GIB and release 3.12.2.

The other adjustments are being tested at the moment in another branch in my fork. "Shadow CI" is still left to do (as is a first iteration to `CONTRIBUTING.md`).

famod commented:
I kicked off a run of the "Shadow CI" in my fork. If all goes as planned I'll update `CONTRIBUTING` and this branch tomorrow and we should be good to go!

famod commented:
**All TODOs are done!** 🎉 (AFAICS)

geoand commented:
EPIC!

famod commented:
FTR, `Quarkus CI incremental` does not include any Windows Job and it has only a single JVM test job (JDK11). Should suffice for the temporary dual/shadow setup, IMO.

famod commented:
Websocket extension renaming got in the way, merge conflict resolved.

famod commented:
It's sad that stef wasn't found, but this has nothing to do with GIB 🤓 
```
Error:  Tests run: 6, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.652 s <<< FAILURE! - in io.quarkus.resteasy.reactive.jsonb.deployment.test.sse.SseTestCase
Error:  io.quarkus.resteasy.reactive.jsonb.deployment.test.sse.SseTestCase.testMultiFromMulti  Time elapsed: 0.013 s  <<< FAILURE!
java.lang.AssertionError: 

Expecting:
  []
to contain exactly (and in same order):
  ["hello", "stef"]
but could not find the following elements:
  ["hello", "stef"]
```

famod commented:
`bom-descriptor-json/pom.xml` keeps getting merge conflicts...

famod committed:
Introduce gitflow-incremental-builder

Relates to #11622

Note: master, release branches (e.g. 1.10) and backport-branches are not built incrementally!

Includes some efficiency teaks to bom-descriptor-json
and a "shadow CI" for a temporary parallel testing phase.

gsmet commented:
OK, let's merge this and see how it goes for a week!

Thanks a lot for your continued work on this, @famod ! Hopefully, it will drastically help with our CI feedback loop!
